### DISCUSSION TITLE
How to prevent double final newlines in `.chezmoitemplates` results

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#3249] [Show and tell]
Many formatters enforce or optionally allow adding a final newline to files.

If you use one of these formatters on a file in your `.chezmoitemplates` directory e.g.:

(`\n` used as a visualization only, not included in the actual file contents)

`.chezmoitemplates/file`:

```text
Hello world
\n
```

and then execute that template inside another template e.g.:

`file.tmpl`

```text
{{ template "file" . }}
\n
```

the resulting file will contain two final newlines:

```text
Hello world
\n
\n
```

If you use the template whitespace control characters in your `.tmpl` file e.g.:

```text
{{ template "file" . -}}
\n
```

the resulting file will have no final newlines:

```text
Hello world
```

Previously, I was working around this by manually removing the final newline from the `.chezmoitemplates` file, which is every bit as tedious as you might think.

If you would like the result to have only a single newline (for clean diffs etc.), you can edit your `.tmpl` file to use [`includeTemplate`](https://www.chezmoi.io/reference/templates/functions/includeTemplate/) in combination with `trim`, and remove the whitespace control characters:

```text
{{ includeTemplate "file" . | trim }}
\n
```

```text
Hello world
\n
```

**Comment by vbrand1984** [comment]
Just add a template comment at the end of the template file:

```
{{- /* This deletes an empty string after the template */ -}}
```
